### PR TITLE
Implement AUTH LOGIN and AUTH PLAIN -- with tests

### PR DESCRIPTION
**Pull request by pepoluan**
## What do these changes do?

**This PR obsoletes #10 and #66**

### MAJOR changes:

* Implement SMTP AUTH with the LOGIN and PLAIN mechanisms.
* Implement extensible AUTH mechanism via the "handler hook" system
* Returns Enhanced Status Code (RFC3463 and RFC5248) for many situations (not yet exhaustive -- focusing mostly on SMTP AUTH and impacted commands)
* Some commands now emits error if AUTH is required but has not happened

### Minor changes:

* No longer targets Python 3.5 because a hard dependency (`atpublic`) requires Python >= 3.6
* Smattering of type hints here there and everywhere

## Are there changes in behavior for the user?

Existing usage should not be impacted.

Users that implemented #66 must implement some changes, though. Namely:

* Change in `SMTP.__init__` kwarg names
* Change in authenticator 'callback' signature

## Major:

#102 

### Minor / related / accidental:

#133 #167 #178 #188 #194 

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] Add a new news fragment into the `CHANGES` folder

### CODE REVIEW

**Review by pepoluan — COMMENTED** [review]
I have confidence in this code. I'd love a **code review**, and if anything is not okay, pray tell.

**Comment by pepoluan** [comment]
Still couldn't find this mythical "CHANGES folder", so I just edited `NEWS.rst` directly.

**Comment by pepoluan** [comment]
> 
> 
> This pull request **fixes 1 alert** when merging [68e006b](https://github.com/aio-libs/aiosmtpd/commit/68e006b4894617c205477054e90e6399f289bd9d) into [97730f3](https://github.com/aio-libs/aiosmtpd/commit/97730f37f4a283b3da3fa3dbf30dd925695fea69) - [view on LGTM.com](https://lgtm.com/projects/g/aio-libs/aiosmtpd/rev/pr-58f41e2a97505beaaea30420010ce175f1fa166a)
> 
> **fixed alerts:**
> 
>     * 1 for Module is imported with 'import' and 'import from'

I seriously have no idea why fixing this Semmle/LGTM alert caused coverage to drop ... but _ONLY_ for Python 3.7.

I do not see the same behavior on my laptop, which runs Windows. And yes, I have run the full tox test set several times.

**Comment by pepoluan** [comment]
Heh, apparently `asyncio` is durn slow on PyPy. It already exhibited that behavior running tox on my WSL. But the test result of Travis confirms it.

I'll remove PyPy from `.travis.yml`.

The still niggling question is why coverage on Python 3.7 on Travis keep "oscillating" between 99% and 100%; and the issue is _always_ between **lines 233~242**.

**Comment by pepoluan** [comment]
PYPY Y U HAZ STRANGE DIR STRUCTURE

**Comment by pepoluan** [comment]
**First of all, about that Travis "Fail"**

Checking the log, it's coverage issue. For building against PyPy3. Which is strange because **all other builds have no coverage problems**.

And since first time my PR went through Travis, if there's coverage problem (outside of strange dir structures -- I'm looking at you PyPy!), it's always this line:

```python
if self.session.ssl is not None:
    return False
```

At this point I'm 99.99% sure that the coverage failure is due to external circumstances, not because of inherent test case inadequacy. So, my suggestion is to **ignore this failure**.

You can check [these screenshots for proof that coverage is 100% when I test](https://imgur.com/a/K3yUc3V) (Imgur album).

I'm done for now. I look forward to any code reviews.

**Comment by pepoluan** [comment]
Gonna close & reopen to trigger travis.

**Comment by pepoluan** [comment]
Let's trigger Travis. I'm 80% certain this time all test passes.

**Comment by pepoluan** [comment]
See? It passes with flying colors! I didn't do anything but close & reopen this PR.

Let's get that Code Review coming!

File: tox.ini
Comment (pepoluan) [inline tox.ini:57]: Reason for ignoring all these:

`E123` -- this is a too-narrow interpretation of [PEP8's "closing brace"](https://www.python.org/dev/peps/pep-0008/#:~:text=The%20closing%20brace) guide.
`E133` -- likewise
`W503` & `W504` -- these are exclusive to one another. If you fulfill W503, you'll violate W504, and vice versa. Also both of them are too-narrow interpretation of [PEP8's "Line break before/after binary operator"](https://www.python.org/dev/peps/pep-0008/#should-a-line-break-before-or-after-a-binary-operator)

(Note, the most recent recommendation is actually `W504`, but that causes a LOT of `qa` issues. Let's fix that in a separate PR)

File: unittest.cfg
Comment (pepoluan) [inline unittest.cfg:4]: `flufl.testing.nose` conflicts with `nose2.plugins.prof`

Because IMHO profiling is more useful than style guide, I sacrificed the former for the latter.

File: unittest.cfg
Comment (pepoluan) [inline unittest.cfg:13]: Actually I'm still failing in capturing the logs so far. But these lines shouldn't mess up with the rest of the tests.

File: unittest.cfg
Comment (pepoluan) [inline unittest.cfg:17]: Combined with the renaming Python script in tox.ini, this allows per-testenv cProfile data.

Also shoving them into `_dynamic` so can easily be cleaned up by removing that single directory.

File: aiosmtpd/smtp.py
Comment (pepoluan) [inline aiosmtpd/smtp.py:50]: Some acrobatics to get type hinting to work nicely.

File: aiosmtpd/smtp.py
Comment (pepoluan) [inline aiosmtpd/smtp.py:176]: Embedded in this block of changes is my attempt to make the code more performant: Instead of keep invoking O(n) `getattr` every time, let's just cache all those `getattr` results in an O(1) dict.

File: aiosmtpd/smtp.py
Comment (pepoluan) [inline aiosmtpd/smtp.py:316]: `str.partition()` and `bytes.partition()` are amazingly fast and result in simpler follow-up code. Let's use them more :-)

Edit: The reason I replaced this was actually an attempt to make `_handle_client` faster. Because in PyPy3 all these for unknown reasons run slower, and enabling PYTHONASYNCIODEBUG flag points me that this method is one of the main reasons for the slowness. This speeds up `_handle_client` enough to greatly reduce the number of "slow callback" warning when testing against PyPy3

File: aiosmtpd/smtp.py
Comment (pepoluan) [inline aiosmtpd/smtp.py:660]: Since we're not using the result of this particular `getattr()` anyways, why not just use `hasattr()`?

File: aiosmtpd/smtp.py
Comment (pepoluan) [inline aiosmtpd/smtp.py:686]: As you can see, caching all those `getattr`s results in a simpler code. Should also be much faster because we can drop the two `str` methods there.

File: aiosmtpd/smtp.py
Comment (pepoluan) [inline aiosmtpd/smtp.py:589]: `not address` already implictly includes `address is None`. Since there is no need to differ if address is None or bool, let's just merge the two into one. Should make `smtp_RCPT` a tad faster.

File: aiosmtpd/smtp.py
Comment (pepoluan) [inline aiosmtpd/smtp.py:666]: Whenever you see `range(len(o))`, 99% of the time it's replaceable with `enumerate(o)`.

File: aiosmtpd/smtp.py
Comment (pepoluan) [inline aiosmtpd/smtp.py:889]: Slicing seems to be quite slow on PyPy3 (subjective feeling).

This is simpler, clearer, and faster. Three bool-ish ops + 1 slicing replaced by an optimized search.

File: aiosmtpd/tests/test_handlers.py
Comment (pepoluan) [inline aiosmtpd/tests/test_handlers.py:660]: See commit message for bca53500

File: aiosmtpd/tests/test_handlers.py
Comment (pepoluan) [inline aiosmtpd/tests/test_handlers.py:702]: See commit message for bca53500

File: aiosmtpd/tests/test_smtp.py
Comment (pepoluan) [inline aiosmtpd/tests/test_smtp.py:17]: These are centralized because they are often used, and if the output ever changes it will be much easier to change the expected output in one place rather than searching this honking big file for all instances of same.

File: aiosmtpd/tests/test_smtp.py
Comment (pepoluan) [inline aiosmtpd/tests/test_smtp.py:33]: Original timeout of `0.1` is too fast for Windows, causing intermittent fails.

By making this into a class attribute it allows easier tuning because changing this automatically adapts the failure case in `TestTimeout.test_timeout`

**Review by pepoluan — COMMENTED** [review]
I have reviewed my code and left comments for some of the less-obvious changes.

File: aiosmtpd/controller.py
Comment (pepoluan) [inline aiosmtpd/controller.py:77]: Similar to #189 

The `pragma: nocover` is _at the moment_ needed to get 100% coverage. I'm trying to implement [`coverage-conditional-plugin`](https://github.com/wemake-services/coverage-conditional-plugin/) but had a glitch with that plugin earlier today. When I fixed the issue with that plugin, I'll replace `nocover` with something better.

**Review by pepoluan — COMMENTED** [review]
I've reviewed changes since I last commitpushed.

Additional comments on the changes have been added.

**Comment by waynew** [comment]
Nice job! I'll take a look at this maybe Monday. I'm only on a phone until then

**Comment by pepoluan** [comment]
*DON'T MERGE THIS YET* ... I have an incoming commit I'm still verifying on my computer. Shouldn't affect coverage or success of test cases, but very likely (I'm more than 90% confident here) solve an intermittent problem I've been experiencing during my test, namely `“Event loop stopped before Future completed.”` error.

**Comment by waynew** [comment]
@pepoluan I converted this to a draft so it can't be merged. I *believe* it will still run all the checks in the interim, though.

File: aiosmtpd/main.py
Comment (pepoluan) [inline aiosmtpd/main.py:150]: This is the point where the error `Event loop stopped before Future completed.` _intermittently_ happened. The root cause was actually in `test_main.py`.

Wrapping some `log.debug` around the scene of the crime, plus a \*very\* eye-catching yellow flag ;-)

File: aiosmtpd/tests/test_main.py
Comment (pepoluan) [inline aiosmtpd/tests/test_main.py:131]: Variable renamed to suppress complaint of shadowing a higher-level `log` variable.

File: aiosmtpd/tests/test_main.py
Comment (pepoluan) [inline aiosmtpd/tests/test_main.py:63]: This is the root cause of the intermittent `Event loop stopped before Future completed.` errors.

Raising the delay to `1.0` from `0.1` and testing the result like 30-40 times seems to indicate that the intermittent problem has been fixed.

File: aiosmtpd/main.py
Comment (pepoluan) [reply aiosmtpd/main.py]: This particular chunk of code has been removed in the final commit because it's no longer needed.

**Review by pepoluan — COMMENTED** [review]
I have reviewed the notable changes between previous commit and this last commit.

File: setup.py
Comment (pepoluan) [inline setup.py:5]: We now have a hard dependency on Python >= 3.6

**Comment by waynew** [comment]
Okay, so I didn't get to this earlier - things went a bit sideways for me at work 🙃 

I will carve out some time either tonight or tomorrow morning to look at this closer!

**Comment by pepoluan** [comment]
> 
> 
> Okay, so I didn't get to this earlier - things went a bit sideways for me at work 🙃
> 
> I will carve out some time either tonight or tomorrow morning to look at this closer!

Feel free to criticize and/or cherry pick 👌

File: .gitignore
Comment (pepoluan) [inline .gitignore:32]: `pyenv` uses this file to limit available Python versions to a project

File: .gitignore
Comment (pepoluan) [inline .gitignore:23]: Suppresses `git status` complaining of new files if tox is cancelled mid-test.

File: .gitignore
Comment (pepoluan) [inline .gitignore:29]: A directory to store dumps of possibly-useful info when running tox's testenvs (e.g., a dump of the currently-seen env vars)

File: aiosmtpd/smtp.py
Comment (pepoluan) [inline aiosmtpd/smtp.py:546]: This is a `WARNING` instead of `INFO` or `DEBUG`, because aborting an `AUTH` mid-mechanism seems ... suspicious.

**Comment by pepoluan** [comment]
Please note that some commits have been cherry-picked into #195 

I strongly recommend that before merging this, #195 should be merged first.

File: .gitignore
Comment (waynew) [reply .gitignore:23]: Would `.coverage*` be a better ignore?

File: .gitignore
Comment (waynew) [reply .gitignore:29]: 👍

File: aiosmtpd/controller.py
Comment (waynew) [reply aiosmtpd/controller.py:14]: I *think* that seems reasonable!

File: aiosmtpd/controller.py
Comment (waynew) [reply aiosmtpd/controller.py:49]: I'd believe it. That is definitely super weird, tho.

File: aiosmtpd/smtp.py
Comment (waynew) [reply aiosmtpd/smtp.py:660]: Can we check the timing of this, over:

```
try:
    return bool(getattr(self, method.__smtp_syntax_when__))
except AttributeError:
    return False
else:
    return True
```

Comment (waynew) [reply aiosmtpd/smtp.py:660]: If there's a significant difference we should do the faster one, but otherwise, I'm cool with this current change.

File: aiosmtpd/tests/test_main.py
Comment (waynew) [reply aiosmtpd/tests/test_main.py:63]: if you change this to something super small like `0.01` can we change it from intermittent to always?

File: tox.ini
Comment (waynew) [reply tox.ini:8]: Will this expose any types of credentials from Travis & co?

File: tox.ini
Comment (waynew) [reply tox.ini:58]: FWIW, I'd be happy to adopt `black` for code formatting, because auto formatting is 👍

**Review by waynew — CHANGES_REQUESTED** [review]
I think this looks great! There are a couple of places where you noted that need some typo fixes. A couple of suggestions/ideas, but overall, 👍  love it!

File: aiosmtpd/controller.py
Comment (pepoluan) [reply aiosmtpd/controller.py:77]: Good idea, I'm putting in the comment now, to go into the next commit round.

File: .gitignore
Comment (pepoluan) [reply .gitignore:23]: That will exclude `.coveragerc` as well, though. The default "partial coverage filename format" is, [according to coverage.py documentation](https://coverage.readthedocs.io/en/coverage-5.3/cmd.html#combining-data-files-coverage-combine), `.coverage.${HOSTNAME}.${PID}.${RANDOM}`.

The two periods there I think was back when the config file was still named `.coverage.ini`, so I had to use two periods to not exclude the config file.

If we want to simplify that exclusion, we still need to leave in at least one period so we won't accidentally exclude `.coveragerc`

File: aiosmtpd/main.py
Comment (pepoluan) [reply aiosmtpd/main.py:160]: I also hate `pragma:nocover`, and actually I have a branch -- still in my laptop, not yet pushed to github -- where I'm dastardly trying to invoke a `KeyboardInterrupt` event.

It's really hard -- if not outright impossible, but I'm working on it -- to do in Windows, though.

So I consider that to be a "placeholder nocover", just like the one in the Python 3.9 compatibility shim (about the `asyncio.all_tasks`).

I'll put in a comment to explain the situation.

File: aiosmtpd/docs/NEWS.rst
Comment (pepoluan) [reply aiosmtpd/docs/NEWS.rst:19]: It's not the docs that mentioned "CHANGES", actually. But the template when creating a new PR.

I see several other PR submitters also got dumbfounded by that checkbox 😁

File: aiosmtpd/smtp.py
Comment (pepoluan) [reply aiosmtpd/smtp.py:155]: Good idea! Why didn't I think of that...

Putting it in the next commit batch as well.

File: aiosmtpd/smtp.py
Comment (pepoluan) [reply aiosmtpd/smtp.py:510]: We can do that for the first five `if`s. Afterwards the `if`s are responding to improper parameter contents, so we shouldn't put them into the `if/elif` chain.

File: aiosmtpd/smtp.py
Comment (pepoluan) [reply aiosmtpd/smtp.py:579]: Good idea. I'm adding a comment in the code to go into the next commit batch.

File: aiosmtpd/smtp.py
Comment (pepoluan) [reply aiosmtpd/smtp.py:660]: Gut feeling says it's about the same 😁. But let me do some `timeit`s and I'll get back to you.

File: tox.ini
Comment (pepoluan) [reply tox.ini:8]: I don't think so ...

That line never show the infodump on-screen; all the dumps go into an 'internal' directory that gets `.gitignore`d and omitted from coverage.

Unless we can access the virtual machine where Travis do their tests, I don't think any Travis info gets leaked.

File: aiosmtpd/smtp.py
Comment (pepoluan) [reply aiosmtpd/smtp.py:166]: Makes sense. Let me add a logging line there to list the 'found' mechanisms.

File: aiosmtpd/smtp.py
Comment (pepoluan) [reply aiosmtpd/smtp.py:510]: Okay, it doesn't result in "too much indentation", which I'm actively avoiding. So I'll implement the change in the next commit batch.

File: aiosmtpd/smtp.py
Comment (pepoluan) [reply aiosmtpd/smtp.py:660]: [Checkout this gist](https://gist.github.com/pepoluan/908996ede1da06c0058061db40d89016)

Surprisingly, the one with `hasattr` is nearly **twice as fast** as the one with `try ... except` 😲

Based on this result, I will use the current implementation with `hasattr`

File: aiosmtpd/tests/test_smtp.py
Comment (pepoluan) [reply aiosmtpd/tests/test_smtp.py:52]: Typo fixed in next commit batch.

File: tox.ini
Comment (pepoluan) [reply tox.ini]: And this is a bit outdated now: I've added `py39` in the list. Works well on my three test platforms.

**Comment by pepoluan** [comment]
I accidentally borked my repo. Please wait while I unbork it

**Comment by pepoluan** [comment]
I don't understand why Github decided that it cannot merge 4 simple changes with the latest commit set, but I've checked the conflicts and all of them can be resoved by accepting the ones from `auth_base`

**Review by pepoluan — COMMENTED** [review]
I believe all requested changes have been implemented, and those that I have implemented, I have marked as "Resolved"

**Comment by pepoluan** [comment]
Gaaah, I'm almost 100% quite certain that it passed `qa` on my laptop /*headdesk/*

Okay will commit a `qa` fix.

File: .gitignore
Comment (pepoluan) [reply .gitignore]: This is merged in #195, so I'm marking as resolved.

File: .coveragerc
Comment (pepoluan) [inline .coveragerc:21]: Custom semi-automatic exclusion of Python 3.9-related lines

File: aiosmtpd/controller.py
Comment (pepoluan) [inline aiosmtpd/controller.py:75]: Yay! We've killed the `nocover` and replaced it with a better "conditional" pragma.

File: tox.ini
Comment (pepoluan) [inline tox.ini:37]: These two lines work together to mark whether the Python interpreter is < 3.9 (`lt`) or >= 3.9 (`ge`); this env var will be used later in `.coveragerc`

File: aiosmtpd/docs/NEWS.rst
Comment (waynew) [reply aiosmtpd/docs/NEWS.rst:28]: Ah. That template should be found in the `.github` folder - we should take some time to fix that!

**Review by waynew — APPROVED** [review]
🚀  Nice!

**Pull request merged**
